src/libpmemfile-posix/pool.c

```diff
@@ -0,0 +1,519 @@
+		if (cred->groups[i] == gid)
+			return true;
+	}
+
+	return false;
+}
+
+/*
+ * can_access -- answers question: "can I access this inode with these
+ * credentials to do specified action?"
+ */
+bool
+can_access(const struct pmemfile_cred *cred, struct inode_perms perms, int acc)
+{
+	pmemfile_mode_t perm = perms.flags & PMEMFILE_ACCESSPERMS;
+	pmemfile_mode_t req = 0;
+	pmemfile_uid_t uid;
+	pmemfile_gid_t gid;
+	int acctype = acc & PFILE_ACCESS_MASK;
+
+	if (acctype == PFILE_USE_FACCESS) {
+		uid = cred->fsuid;
+		gid = cred->fsgid;
+	} else if (acctype == PFILE_USE_EACCESS) {
+		uid = cred->euid;
+		gid = cred->egid;
+	} else if (acctype == PFILE_USE_RACCESS) {
+		uid = cred->ruid;
+		gid = cred->rgid;
+	} else {
+		return false;
+	}
+
+	if (perms.uid == uid) {
+		if (acc & PFILE_WANT_READ)
+			req |=  PMEMFILE_S_IRUSR;
+		if (acc & PFILE_WANT_WRITE)
+			req |=  PMEMFILE_S_IWUSR;
+		if (acc & PFILE_WANT_EXECUTE)
+			req |=  PMEMFILE_S_IXUSR;
+	} else if (perms.gid == gid || gid_in_list(cred, perms.gid)) {
+		if (acc & PFILE_WANT_READ)
+			req |=  PMEMFILE_S_IRGRP;
+		if (acc & PFILE_WANT_WRITE)
+			req |=  PMEMFILE_S_IWGRP;
+		if (acc & PFILE_WANT_EXECUTE)
+			req |=  PMEMFILE_S_IXGRP;
+	} else {
+		if (acc & PFILE_WANT_READ)
+			req |=  PMEMFILE_S_IROTH;
+		if (acc & PFILE_WANT_WRITE)
+			req |=  PMEMFILE_S_IWOTH;
+		if (acc & PFILE_WANT_EXECUTE)
+			req |=  PMEMFILE_S_IXOTH;
+	}
+
+	return ((perm & req) == req);
+}
+
+/*
+ * vinode_can_access -- wrapper around "can_access" that deals with locked
+ * vinode
+ */
+bool
+_vinode_can_access(const struct pmemfile_cred *cred,
+		struct pmemfile_vinode *vinode, int acc)
+{
+	struct inode_perms inode_perms = _vinode_get_perms(vinode);
+
+	return can_access(cred, inode_perms, acc);
+}
+
+/*
+ * vinode_can_access -- wrapper around "can_access" that deals with unlocked
+ * vinode
+ */
+bool
+vinode_can_access(const struct pmemfile_cred *cred,
+		struct pmemfile_vinode *vinode, int acc)
+{
+	struct inode_perms inode_perms = vinode_get_perms(vinode);
+
+	return can_access(cred, inode_perms, acc);
+}
+
+/*
+ * copy_cred -- copies credentials
+ */
+static int
+copy_cred(struct pmemfile_cred *dst_cred, const struct pmemfile_cred *src_cred)
+{
+	dst_cred->ruid = src_cred->ruid;
+	dst_cred->rgid = src_cred->rgid;
+
+	dst_cred->euid = src_cred->euid;
+	dst_cred->egid = src_cred->egid;
+
+	dst_cred->fsuid = src_cred->fsuid;
+	dst_cred->fsgid = src_cred->fsgid;
+
+	dst_cred->caps = src_cred->caps;
+	dst_cred->groupsnum = src_cred->groupsnum;
+	if (dst_cred->groupsnum) {
+		dst_cred->groups = malloc(dst_cred->groupsnum *
+				sizeof(dst_cred->groups[0]));
+		if (!dst_cred->groups)
+			return -1;
+		memcpy(dst_cred->groups, src_cred->groups, dst_cred->groupsnum *
+				sizeof(dst_cred->groups[0]));
+	} else {
+		dst_cred->groups = NULL;
+	}
+
+	return 0;
+}
+
+/*
+ * get_cred -- gets current credentials in a safe way
+ */
+int
+get_cred(PMEMfilepool *pfp, struct pmemfile_cred *cred)
```

Comment:
That's exactly what I was talking about previously. Why don't we use this everywhere?